File: Cpp/Maximum_Frequency_of_an_Element_After_Performing_Operations_I.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k, int numOperations) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        long long k64 = k;
        vector<long long> candidates;
        for (int x : nums) {
            candidates.push_back(x);
            candidates.push_back(x - k);
            candidates.push_back(x + k);
        }
        sort(candidates.begin(), candidates.end());
        candidates.erase(unique(candidates.begin(), candidates.end()),
                         candidates.end());

        int ans = 0;
        for (long long t : candidates) {
            long long left = t - k64;
            long long right = t + k64;
            int lo = lower_bound(nums.begin(), nums.end(), left) - nums.begin();
            int hi =
                upper_bound(nums.begin(), nums.end(), right) - nums.begin();
            int cover = hi - lo;
            int base_lo =
                lower_bound(nums.begin(), nums.end(), t) - nums.begin();
            int base_hi =
                upper_bound(nums.begin(), nums.end(), t) - nums.begin();
            int base = base_hi - base_lo;
            int candidates_count = cover - base;
            int freq = base + min(candidates_count, numOperations);
            ans = max(ans, freq);
        }
        return ans;
    }
};

File: Rust/Maximum_Frequency_of_an_Element_After_Performing_Operations_I.rs
use std::cmp::min;

impl Solution {
    pub fn max_frequency(nums: Vec<i32>, k: i32, num_operations: i32) -> i32 {
        let mut nums = nums;
        nums.sort();
        let n = nums.len();
        let k64 = k as i64;
        let mut candidates = Vec::new();
        for &x in &nums {
            candidates.push(x);
            candidates.push(x - k);
            candidates.push(x + k);
        }
        candidates.sort();
        candidates.dedup();

        let mut ans = 0;
        for &t in &candidates {
            let t64 = t as i64;
            let left = t64 - k64;
            let right = t64 + k64;
            let lo = nums.partition_point(|&val| (val as i64) < left);
            let hi = nums.partition_point(|&val| (val as i64) <= right);
            let cover = (hi - lo) as i32;
            let base_lo = nums.partition_point(|&val| (val as i64) < t64);
            let base_hi = nums.partition_point(|&val| (val as i64) <= t64);
            let base = (base_hi - base_lo) as i32;
            let candidates_count = cover - base;
            let freq = base + min(candidates_count, num_operations);
            ans = ans.max(freq);
        }
        ans
    }
}
